pkg/operators/ebpf/tracer.go

```diff
@@ -0,0 +1,207 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ebpfoperator
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/btf"
+	"github.com/cilium/ebpf/perf"
+	"github.com/cilium/ebpf/ringbuf"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	metadatav1 "github.com/inspektor-gadget/inspektor-gadget/pkg/metadata/v1"
+)
+
+type Tracer struct {
+	metadatav1.Tracer
+
+	ds       datasource.DataSource
+	accessor datasource.FieldAccessor
+
+	mapType       ebpf.MapType
+	ringbufReader *ringbuf.Reader
+	perfReader    *perf.Reader
+}
+
+func validateTraceMap(traceMap *ebpf.MapSpec) error {
+	if traceMap.Type != ebpf.RingBuf && traceMap.Type != ebpf.PerfEventArray {
+		return fmt.Errorf("map %q has a wrong type, expected: ringbuf or perf event array, got: %s",
+			traceMap.Name, traceMap.Type.String())
+	}
+	return nil
+}
+
+func (i *ebpfInstance) populateTracer(t btf.Type, varName string) error {
```

Comment:
When building, users can use the `--update-metadata` flag to generate the metadata from scratch or keep it updated. How can the build command reuse this logic to generate the metadata? I don't want to have this logic repeated in both places.

Reply:
I think `GetGadgetInfo()` could be used to do that later on (might need some minor modifications).

Reply:
Aaahh I thought the `GetGadgetInfo()` grpc call would be gone after removing the `pkg/gadgets/run` in this PR.